src/listener.js

```diff
@@ -2,156 +2,226 @@
-        log('connection closed before p2p connection made')
-      } else {
-        log('error interpreting incoming p2p connection')
-      }
+  /**
+   * Closes all open servers
+   * @param {object} options
+   * @param {number} options.timeout how long before closure is forced, defaults to 2000 ms
+   * @returns {Promise}
+   */
+  close (options = {}) {
+    if (!this._isListening()) {
       return
     }
 
-    log('new connection', addr.toString())
+    // Close all running servers in parallel
+    return Promise.all(
+      [...this._servers].map(server => {
+        return new Promise((resolve, reject) => {
+          const start = Date.now()
+
+          // Attempt to stop the server. If it takes longer than the timeout,
+          // destroy all the underlying sockets manually.
+          const timeout = setTimeout(() => {
+            log('Timeout closing server after %dms, destroying connections manually', Date.now() - start)
+            resolve()
+          }, options.timeout || c.CLOSE_TIMEOUT)
+
+          server.once('close', () => {
+            clearTimeout(timeout)
+            this._servers.delete(server)
+          })
 
-    const s = new Libp2pSocket(socket, addr)
-    trackSocket(server, socket)
+          server.close((err) => err ? reject(err) : resolve())
+        })
+      })
+    ).then(() => {
+      this.__connections.forEach((connection, key) => {
+        log('destroying %s', key)
+        connection.destroy()
+      })
+      this.__connections.clear()
+      this._servers.clear()
+    })
+  }
 
-    handler && handler(s)
-    listener.emit('connection', s)
-  })
+  /**
+   * Creates servers listening on the given `addrs`
+   * @async
+   * @param {Array<Multiaddr>} addrs
+   */
+  async listen (addrs) {
+    addrs = Array.isArray(addrs) ? addrs : [addrs]
+
+    let listeners = []
+    let errors = []
+
+    // Filter out duplicate ports, unless it's port 0
+    addrs = uniqueBy(addrs, (addr) => {
+      const port = Number(addr.toOptions().port)
+      return isNaN(port) || port === 0 ? addr.toString() : port
+    })
 
-  server.on('listening', () => listener.emit('listening'))
-  server.on('error', (err) => listener.emit('error', err))
-  server.on('close', () => listener.emit('close'))
+    for (const ma of addrs) {
+      const lOpts = ma.toOptions()
 
-  // Keep track of open connections to destroy in case of timeout
-  server.__connections = {}
+      listeners.push(
+        new Promise((resolve) => {
+          const server = net.createServer(this._onSocket.bind(this))
+          this._servers.add(server)
 
-  listener.close = (options = {}) => {
-    if (!server.listening) {
-      return
-    }
+          server.on('listening', () => this.emit('listening'))
+          server.on('close', () => {
+            // only emit if we're not listening
+            if (!this._isListening()) {
+              this.emit('close')
+            }
+            this._servers.delete(server)
```

Comment:
I think we may need to remove the event listeners in order for the server objects to get GCed

Reply:
Added removeAllListeners to this logic.